query-algebrizer/src/clauses/where_fn.rs

```diff
@@ -243,6 +252,172 @@ impl ConjoiningClauses {
+                TypedValue::typed_string(s.as_str())
+            },
+            _ => bail!(ErrorKind::InvalidArgument(where_fn.operator.clone(), "string".into(), 2)),
+        };
+
+        // TODO: should we build the FQA in ::Matches, preventing nonsense like matching on ::Rowid?
+        let constraint = ColumnConstraint::Matches(QualifiedAlias(fulltext_values_alias.clone(), Column::Fulltext(FulltextColumn::Text)),
+                                                   QueryValue::TypedValue(search));
+        self.wheres.add_intersection(constraint);
+
+        if let VariableOrPlaceholder::Variable(ref var) = bindings[0] {
+            if self.bound_value(&var).is_some() || self.input_variables.contains(&var) {
+                bail!(ErrorKind::InvalidBinding(where_fn.operator.clone(), BindingError::BoundInputVariable));
+            }
+            self.constrain_var_to_type(var.clone(), ValueType::Ref);
+
+            let entity_alias = QualifiedAlias(datoms_table_alias.clone(), Column::Fixed(DatomsColumn::Entity));
+            self.column_bindings.entry(var.clone()).or_insert(vec![]).push(entity_alias);
+        }
+
+        if let VariableOrPlaceholder::Variable(ref var) = bindings[1] {
+            if self.bound_value(&var).is_some() || self.input_variables.contains(&var) {
+                bail!(ErrorKind::InvalidBinding(where_fn.operator.clone(), BindingError::BoundInputVariable));
+            }
+            self.constrain_var_to_type(var.clone(), ValueType::String);
+
+            let value_alias = QualifiedAlias(fulltext_values_alias.clone(), Column::Fulltext(FulltextColumn::Text));
+            self.column_bindings.entry(var.clone()).or_insert(vec![]).push(value_alias);
+        }
+
+        if let VariableOrPlaceholder::Variable(ref var) = bindings[2] {
+            if self.bound_value(&var).is_some() || self.input_variables.contains(&var) {
+                bail!(ErrorKind::InvalidBinding(where_fn.operator.clone(), BindingError::BoundInputVariable));
+            }
+            self.constrain_var_to_type(var.clone(), ValueType::Ref);
+
+            let tx_alias = QualifiedAlias(datoms_table_alias.clone(), Column::Fixed(DatomsColumn::Tx));
+            self.column_bindings.entry(var.clone()).or_insert(vec![]).push(tx_alias);
+        }
+
+        if let VariableOrPlaceholder::Variable(ref var) = bindings[3] {
+            if self.bound_value(&var).is_some() || self.input_variables.contains(&var) {
+                bail!(ErrorKind::InvalidBinding(where_fn.operator.clone(), BindingError::BoundInputVariable));
+            }
+            self.constrain_var_to_type(var.clone(), ValueType::Double);
+
+            // Right now we don't support binding a column to a constant value.  Therefore, we
+            // introduce a computed table with exactly the constant value we require.  See the
+            // translate tests for some edge cases in this space.
+            // TODO: optimize this by binding the value, like:
```

Comment:
Yup. But as with `ground`, doing so requires awareness of whether there are any existing patterns. In that case you can spit out a simple scalar `FROM` like `(SELECT 0.0 AS ?match) AS c00` so that it'll join with existing columns.